ISDK-1523: Add ARKit iOS 11 Example

ceaglest opened the pull request:
This PR adds an example which uses ARKit with Twilio Video. This project was originally authored by Lizzie Siegle as a part of a blog post. I took the original source code and integrated it into the QuickStart, at a high level:

* Implement protocols in extensions
* Implement full TVIVideoCapturer lifecycle.
* Use common project hierarchy, and project setup from QS.
* Add a project level README.md and integrate into the top level README.md.
* Use copying of CGImages rather than drawing using CGContext. This is 1-2 orders of magnitude faster.
* Don't use TVIScreenCapturer.
* Fix BGRA vs ARGB format description.
* Use Swift 3.2 instead of Swift 4.0 due to issues with CocoaPods integration.
* Removed dead code, comments and reformatted some sections.

### Discussion

This example doesn't have any explicit UI to connect to a Room. The connection behavior is automatic and the name hardcoded to `arkit`. Also there is no UI to display remote Participants (but you can hear them). One thing that might be nice is an explicit error dialog when connecting to the Room fails. Errors and important events are logged to the console for now.

As for CocoaPods and Swift language versions, I ran into a the issue noted in: https://github.com/CocoaPods/CocoaPods/issues/5864. I didn't dig too deep into this one, and instead opted to use Swift 3.2 to be consistent with the other examples.

@elizabethsiegle please review this PR as well. I wasn't able to add you to the reviewers section since it's restricted to the Twilio organization.

ceaglest commented on ARKitExample/README.md:
```diff
@@ -0,0 +1,19 @@
+# Twilio Video ARKit Example
+
+The project demonstrates how to use Twilio's Programmable Video SDK to stream an augmented reality scene created with ARKit and SceneKit. This example was originally provided as part of a blog post that you can find [here]().
```
Needs a link to the blog post.

ceaglest commented on ARKitExample/ViewController.swift:
```diff
@@ -0,0 +1,204 @@
+
+    func room(_ room: TVIRoom, didFailToConnectWithError error: Error) {
+        print("Failed to connect to a Room: \(error).")
+    }
```
This is probably worthy of a UIAlertView since developers may forget to add a token.

paynerc replied:
Yes, I agree. We should definitely display something to alert the user.

ceaglest replied:
👍 I think this is the minimal UI we can get away with before merging the example.

ceaglest commented on README.md:
```diff
@@ -76,8 +76,9 @@ Note: If you have an iOS device, you can now run apps from Xcode on your device
+- [ARKit](ARKitExample) - Captures augmented reality content with `ARKit` and uses `SceneKit` to place objects in the virtual scene. The resulting rendering is previewed locally, and shared in a `TVIRoom` via a custom `TVIVideoCapturer` source. This app requires Xcode 9 & iOS 11.
 - [Screen Capturer](ScreenCapturerExample) - Shows how to use `TVIScreenCapturer` to capture the contents of a `UIView`, and how a custom `TVIVideoCapturer` can be implemented to do the same.
-- [Video CallKit](VideoCallKitQuickStart) - Shows how to use Twilio Video with the [CallKit framework](https://developer.apple.com/reference/callkit).
+- [Video CallKit](VideoCallKitQuickStart) - Shows how to use Twilio Video with the [CallKit](https://developer.apple.com/reference/callkit)  framework.
```
Extra space.

ceaglest commented on Podfile:
```diff
@@ -1,20 +1,27 @@
   end
 
   target 'ScreenCapturerExample' do
+    platform :ios, '8.1'
```
Specifying the platform at the target level seemed like the clearest thing to do.

elizabethsiegle commented:
Reviewed! 👍

ceaglest commented:
@elizabethsiegle lets make this QuickStart repo the single source of code for the ARKit example. Your blog post can link directly reference the code added in this PR.

elizabethsiegle commented:
@ceaglest linking now! Updated code in post re. tviscreencapturer too

idelgado commented on ARKitExample/README.md:
```diff
@@ -0,0 +1,19 @@
+
+This example requires Xcode 9.0, and the iOS 11.0 SDK. An iOS device with an A9 CPU or greater is needed for ARKit to function properly.
+
+### Usage
```
I think this detail is important to add to the README.

> For this to work properly **you need to generate a new access token with a different identity** and connect to the same room Arkit using the quickstart app.

ceaglest replied:
Fair enough. Let me take one more pass on this section.

paynerc commented on ARKitExample.xcodeproj/project.xcworkspace/contents.xcworkspacedata:
```diff
@@ -0,0 +1,7 @@
+<Workspace
+   version = "1.0">
+   <FileRef
+      location = "self:/Users/ceagleston/Development/video-quickstart-swift/ARKitExample/ARKitExample.xcodeproj">
```
Uh.. You might want to revisit this, @ceaglest

ceaglest replied:
Whoops. Let me clean this up. 👍

paynerc commented on ARKitExample/AppDelegate.swift:
```diff
@@ -0,0 +1,45 @@
+//  AppDelegate.swift
+//  ARKitExample
+//
+//  Created by Lizzie Siegle on 8/10/17.
```
File header for the quickstarts shouldn't have the author line.

ceaglest replied:
Normally we do not, I was wondering if we should keep this considering Lizzie originally wrote the example and she isn't a part of the iOS SDK team.

paynerc commented on ARKitExample/ViewController.swift:
```diff
@@ -0,0 +1,204 @@
+//  ViewController.swift
+//  ARKitExample
+//
+//  Created by Lizzie Siegle on 8/10/17.
```
Remove author line

ceaglest commented on ARKitExample/ViewController.swift:
```diff
@@ -0,0 +1,204 @@
+        // We only support the single capture format that our ARSession provides.
+        // Since ARSession doesn't seem to give us any way to query its output format, we will assume 1280x720x30
+        let format = TVIVideoFormat.init()
+        format.dimensions = CMVideoDimensions.init(width: 1280, height: 720)
```
This is kind of sketchy. We definitely aren't getting 1280x720 by rendering at 1x resolution. Effectively our capturer only supports a single format, but I could probably make a better guess than this.

paynerc commented on ARKitExample/ViewController.swift:
```diff
@@ -0,0 +1,204 @@
+        self.audioTrack = TVILocalAudioTrack.init()
+
+        let options = TVIConnectOptions.init(token: accessToken, block: {(builder: TVIConnectOptionsBuilder) -> Void in
+            builder.videoTracks = [self.videoTrack!]
```
I don't particularly care for the forced unwrap here.

To be more idiomatic Swift, you could do something like:

```
        if let videoTrack = self.videoTrack {
            builder.videoTracks = [videoTrack]
        }

        if let audioTrack = self.audioTrack {
            builder.audioTracks = [audioTrack]
        }
````

paynerc commented:
Do you want to add the ARKitExample to the `VideoQuickStart.xcworkspace` as well?

ceaglest commented:
@paynerc I'm expecting CocoaPods to generate the `.xcworkspace`, and there is no manual variant of it. Is there something I'm missing in the `Podfile`? It worked when tested locally.

ceaglest commented:
@paynerc @idelgado I've addressed your feedback. Outstanding items include:

1. Author line - Mention in README, and remove in the source code?
2. Final link to the blog post - Could be added in another PR if we'd like.

paynerc commented on ARKitExample/ViewController.swift:
```diff
@@ -183,6 +173,17 @@ extension ViewController: TVIVideoCapturer {
+        format.dimensions = CMVideoDimensions.init(width: Int32(viewSize.width), height: Int32(viewSize.height))
+        format.frameRate = UInt(sceneView.preferredFramesPerSecond)
+        format.pixelFormat = TVIPixelFormat.format32BGRA
+        return []
```
You aren't returning anything here?

paynerc commented on ARKitExample/ViewController.swift:
```diff
@@ -0,0 +1,219 @@
+                                                  &pixelBuffer)
+
+        if (status != kCVReturnSuccess) {
+            return nil as CVPixelBuffer!;
```
This seems problematic? We are saying that this method returns a `CVPixelBuffer`, but in this pathway we are returning `nil`? I am surprised that the compiler doesn't get angry about this. Seems like make we should just make the return value an optional and not do the force unwrapping.

ceaglest replied:
Good point, I think making the return optional is the right way to go. Silly copy/paste of similar code from ScreenCapturerExample which didn't have the same nullability requirements.

paynerc commented on ARKitExample/ViewController.swift:
```diff
@@ -0,0 +1,219 @@
+            return nil as CVPixelBuffer!;
+        }
+
+        return pixelBuffer!
```
See above

paynerc commented:
WRT the author line: I think a mention in the `README` would probably be sufficient? And the code would be consistent across all of the example apps?

ceaglest commented:
@paynerc I addressed your feedback, once I have your +1 I think we are good to merge, and add a link to the live blog post when its ready.

ceaglest commented on ARKitExample/ViewController.swift:
```diff
@@ -0,0 +1,218 @@
+        let format = TVIVideoFormat.init()
+        format.frameRate = UInt(sceneView.preferredFramesPerSecond)
+        format.pixelFormat = TVIPixelFormat.format32BGRA
+        return [format]
```
I was getting an invalid constraints error when specifying the screen size as the dimensions for the TVIVideoFormat. I will file a bug and look into this, but I don't think we should let it gate merging the PR.

paynerc commented on ARKitExample/ViewController.swift:
```diff
@@ -89,17 +89,19 @@ class ViewController: UIViewController {
-                                   buffer: pixelBuffer,
-                                   orientation: TVIVideoOrientation.up)
-        self.consumer?.consumeCapturedFrame(self.frame!)
+        if (pixelBuffer != nil) {
```
We really should use if/let bindings here instead of doing a nil check and force unwrapping in the block. But it's something that we can address in the future.

paynerc commented on ARKitExample/ViewController.swift:
```diff
@@ -89,17 +89,19 @@ class ViewController: UIViewController {
      * This pretty fast but still wasteful, it would be nicer to wrap the CGImage and use its CGDataProvider directly.
      **/
-    func copyPixelbufferFromCGImageProvider(image: CGImage) -> CVPixelBuffer {
+    func copyPixelbufferFromCGImageProvider(image: CGImage) -> CVPixelBuffer? {
```
👍

ceaglest commented:
Thanks for the reviews everyone! I'll submit another PR once the blog post is live with an updated link.

elizabethsiegle commented:
Is blog post good to go on your end? Also waiting for final confirmation from Brent, my teammate doing tech review
